rust/src/storage/s3/store.rs

```diff
@@ -0,0 +1,734 @@
+    model::Model,
+};
+use futures::ready;
+use rusoto_core::{credential::ProvideAwsCredentials, HttpClient, Region, RusotoError};
+use rusoto_s3::{
+    CreateBucketError,
+    CreateBucketOutput,
+    CreateBucketRequest,
+    Delete,
+    DeleteObjectsError,
+    DeleteObjectsOutput,
+    DeleteObjectsRequest,
+    GetObjectError,
+    GetObjectOutput,
+    GetObjectRequest,
+    ListObjectsV2Error,
+    ListObjectsV2Output,
+    ListObjectsV2Request,
+    ObjectIdentifier,
+    PutObjectError,
+    PutObjectOutput,
+    PutObjectRequest,
+    S3Client,
+    StreamingBody,
+    S3,
+};
+use std::{
+    collections::HashMap,
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use thiserror::Error;
+use tokio::{io::AsyncReadExt, stream::Stream};
+
+#[derive(Debug, Error)]
+pub enum StorageError {
+    #[error("S3Error")]
+    S3Error,
+}
+
+#[derive(Clone)]
+struct Buckets(HashMap<&'static str, &'static str>);
+
+impl Buckets {
+    fn new() -> Self {
+        let mut names = HashMap::new();
+        names.insert("masks", "masks");
+        names.insert("masked-models", "masked-models");
+        names.insert("global-model", "global-model");
+        names.insert("global-masked-model", "global-masked-model");
+        Self(names)
+    }
+
+    fn masks(&self) -> &'static str {
+        self.0.get("masks").unwrap()
+    }
+
+    fn masked_models(&self) -> &'static str {
+        self.0.get("masked-models").unwrap()
+    }
+
+    fn global_model(&self) -> &'static str {
+        self.0.get("global-model").unwrap()
+    }
+
+    fn global_masked_model(&self) -> &'static str {
+        self.0.get("global-masked-model").unwrap()
+    }
+
+    fn names(&self) -> impl Iterator<Item = &&str> {
+        self.0.values()
+    }
+}
+
+#[derive(Clone)]
+pub struct S3Store {
+    s3_client: S3Client,
+    buckets: Buckets,
+}
+
+// API
+impl S3Store {
+    /// Create a new S3 store. The store creates and maintains 4 different buckets.
+    /// See [`Buckets`] for more information.
+    ///
+    /// The [`S3Store`] can also be used together with Minio via a custom region:
+    /// ```text
+    /// let region = Region::Custom {
+    ///    name: String::from("eu-east-3"),
+    ///    endpoint: String::from("http://127.0.0.1:9000"), // URL of minio
+    /// };
+    ///
+    /// let store = S3Store::new(region);
+    /// ```
+    pub fn new<P>(region: Region, credentials_provider: P) -> Self
+    where
+        P: ProvideAwsCredentials + Send + Sync + 'static,
+    {
+        let dispatcher = HttpClient::new().expect("failed to create request dispatcher");
+        Self {
+            s3_client: S3Client::new_with(dispatcher, credentials_provider, region),
+            buckets: Buckets::new(),
+        }
+    }
+
+    /// Upload a [`Mask`].
+    pub async fn upload_mask(&self, key: &str, mask: &Mask) -> Result<(), StorageError> {
+        match self
+            .upload(self.buckets.masks(), key, mask.serialize())
```

Comment:
The masks and models are pretty big, and the route I'm taking right now in the masking refactoring is to store them as `Vec<BigUint>` (for masked models) and `Vec<Ratio<BigInt>>` (for models), which will make them even bigger. So (de)serialization might take a non-negligeable amount of time, to the point that it might be worth considering using `block_on` for these operations.

Do you know if rusoto has a streaming API? Like can we upload a `tokio::Stream<Item=u8>`? If so I think it's worth investigating that area as well.

This can't be addressed now of course since we need benchmarks. Just something to keep in mind for the future.

Reply:
The value of the `body` in the `PutObjectRequest` is a [`StreamingBody`](https://docs.rs/rusoto_s3/0.43.0/rusoto_s3/struct.PutObjectRequest.html#structfield.body) which is an alias for [`ByteStream`](https://docs.rs/rusoto_core/0.43.0/rusoto_core/struct.ByteStream.html). I think this is what you mean right?

Reply:
Yeah something like this. But `ByteStream` still requires the entire `Vec<u8>`. I was hoping we could serialize on the fly instead.